### PR TITLE
Fixed #20892 -- Allowed setting memcached Client options through CACHES 'OPTIONS'.

### PR DESCRIPTION
**Pull request by Fiedzia**
Allow to set memcached client options
Ticket https://code.djangoproject.com/ticket/20892

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, add to whitelist.

File: django/core/cache/backends/memcached.py [hunk 1]
```diff
@@ -32,7 +32,12 @@ def _cache(self):
         Implements transparent thread-safe access to a memcached client.
         """
         if getattr(self, '_client', None) is None:
-            self._client = self._lib.Client(self._servers)
+            if self._options and 'CLIENT_OPTIONS' in self._options:
```
Comment (timgraham) [inline django/core/cache/backends/memcached.py]: You could avoid the two branches with:
 ```
self._options = params.get('OPTIONS') or {}
self._options.setdefault('CLIENT_OPTIONS', {})
```
on line 27.

File: docs/topics/cache.txt [hunk 2]
```diff
@@ -123,6 +123,27 @@ When using the ``pylibmc`` binding, do not include the ``unix:/`` prefix::
         }
     }
 
+You can set CLIENT_OPTIONS in OPTIONS setting to pass additional parameter
```
Comment (timgraham) [inline docs/topics/cache.txt]: use double backticks around code stuff like CLIENT_OPTIONS for proper formatting.

File: docs/topics/cache.txt [hunk 3]
```diff
@@ -123,6 +123,27 @@ When using the ``pylibmc`` binding, do not include the ``unix:/`` prefix::
...
+to memcached client. For example::
+
+    CACHES = {
+        'default': {
+            'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
+            'LOCATION': '127.0.0.1:11211',
+            'OPTIONS': {
+                'CLIENT_OPTIONS': {
+                    'socket_timeout': 1
```
Comment (timgraham) [inline docs/topics/cache.txt]: add trailing comma

File: docs/topics/cache.txt [hunk 4]
```diff
@@ -123,6 +123,27 @@ When using the ``pylibmc`` binding, do not include the ``unix:/`` prefix::
...
+                }
+            }
+        }
+    }
+
+will pass socket_timeout option to memcached client instance, limiting time
+spent waiting for connection with potentially unavailable memcached server
+to 1 second instead of default 3.
+Consult client ``__init__`` method to get list of available options
```
Comment (timgraham) [inline docs/topics/cache.txt]: If this is meant as a separate paragraph, include a blank line above it, otherwise flow the sentence to the previous line.

**Comment by timgraham** [comment]
Please check the [patch review checklist](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) carefully. A few requirements such as release notes and a `.. versionchanged:: 1.10` annotation are missing.

**Comment by Fiedzia** [comment]
Shouldn't it be versionadded insted of versionchanged?

**Comment by timgraham** [comment]
Please check the tests with pylibmc (it's not working):
```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.memcached.PyLibMCCache',
        'LOCATION': '127.0.0.1:11211',
    },
}
```
Here are a few other edits: http://dpaste.com/2RB14P8

**Comment by timgraham** [comment]
You can squash commits and use the commit message from the pull request title. Uncheck "Patch needs improvement" on the ticket when you update it, thanks!

**Comment by Fiedzia** [comment]
Right, pylibmc is a bit different, since it passses OPTIONS to client.behaviors. 
This seems wrong to me, as it closes the way use options for backend options, instead of just subset of client options.
I could solve this in following ways:
1. Make a copy of OPTIONS and pass them with CLIENT_OPTIONS removed.
2. Move CLIENT_OPTIONS  out of OPTIONS.
3. Change pylibmc to not pass OPTIONS and instead use CLIENT_OPTIONS to set client behaviors. That's sound like the right thing to do to me, but it may break people's code.

Does anyone has an opinion on that? If not I'd go with 2 or 3.

**Comment by timgraham** [comment]
Yes, I think we should aim for #3 but with a deprecation path. If `CLIENT_OPTIONS` doesn't appear in `OPTIONS`, then continue the old behavior but raise a deprecation warning about it. Maybe shortening `CLIENT_OPTIONS` to `CLIENT` would be better (since we're already in `OPTIONS`). 

The other option could be to omit `CLIENT` entirely and pass `OPTIONS` to the `Client`. This seems most consistent with [the documentation](https://docs.djangoproject.com/en/dev/topics/cache/#cache-arguments) which says, "cache backends backed by a third-party library will pass their options directly to the underlying cache library." Of course this will preclude adding any more OPTIONS for non-Client features that might arise in the future.

Probably we should raise this on the django-developers mailing list after the two of us (and anyone else reading here) come to a consensus.

**Comment by timgraham** [comment]
Any response to the last point? Will you continue work on this? Thanks!

**Comment by timgraham** [comment]
Closing due to inactivity. Please send a new PR if you continue this.

**Comment by edmorley** [comment]
> The other option could be to omit CLIENT entirely and pass OPTIONS to the Client

I think this is the cleanest/most intuitive option, and also happens to have been proposed in a few previous discussions:
https://groups.google.com/d/msg/django-developers/pISp20wuu0E/tamJ1h8zCzsJ
https://code.djangoproject.com/ticket/15815#comment:3

This would mean for pylibmc behaviors are then set like so:

```
CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.memcached.PyLibMCCache',
        'LOCATION': '127.0.0.1:11211',
        'OPTIONS': {
            'binary': True,
            'behaviors': {
                'ketama': True,
                'tcp_nodelay': True
            }
        }
    },
}
```

If we're happy to go with that approach, the only question is how to best handle deprecation warnings.

We can't just check for the absence of `...['OPTIONS']['behaviors']`, since it's feasible someone may actually be wanting to set a top level parameter (eg `binary=True`) but not set any `behaviors`.

As such, I think the next best approach is something like (for pylibmc's backend only):
* check whether there are any keys in `OPTIONS` that aren't `('username', 'password', 'binary', 'behaviors')` (ie the keyword arguments to the pylibmc Client constructor; it's a shorter list than checking for the inverse)
* if there are, output a deprecation warning, pop them from `OPTIONS`, and instead set them under `...['OPTIONS']['behaviors']` instead

If that sounds reasonable to you @timgraham, I'll open a PR :-)

**Comment by timgraham** [comment]
It does.

**Comment by edmorley** [comment]
Fixed in PR #7160.

**Pull request closed without merging**